### ISSUE TITLE
When a tensor is square (NxN), what can I do to invert it without writing my own implementation?

### ISSUE THREAD

**Issue by ArthurHypersphere** [elixir-nx/nx#268]

**Comment by elbow-jason** [comment]
The word "invert" here is ambiguous. Can you give an example of what your square tensor looks like before and what your square tensor looks like after it is inverted?

**Comment by ArthurHypersphere** [comment]
I mean it in the sense of an invertible matrix, i.e.

AB == BA == (identity matrix of the same size)

**Comment by seanmor5** [comment]
I don't think it's possible with our current API. In NumPy you would do something like:

```python
a = np.array([[2, 1, 2], [3, 2, 3], [4, 5, 6]], dtype=np.float32)
np.linalg.inv(a)
```

Given the expected API after we resolve #157,  under the hood, I think it'd be something like:

```elixir
defn invert_matrix(a) do
  solve(a, Nx.eye(a)) # solve `a` for the identity matrix
end

defn solve(a) do
  lu = Nx.lu(a)
  Nx.triangular_solve(lower)
end
```

I don't know how long it's going to be until we are able to getting around to tackling those implementations. @polvalente has been pretty active in handling those. He might be able to help more

**Comment by polvalente** [comment]
@ArthurHypersphere Depending on your use-case, QR-factorization could remove the need to invert a matrix. 

If you're trying to find `x` for `Ax = b`, if you have `A = QR`, where Q is  orthonormal and R is upper triangular, you can turn the problem into `Rx = transpose(Q)b`, which is easily solvable through back substitution. QR decomposition is being implemented in #265 and I think there'll be a pull request for `triangular_solve` coming soon. If not, an implementation would be something akin to:

```elixir

for row <- n-1..0, col <- n-1..row do
```

**Comment by polvalente** [comment]
@ArthurHypersphere Depending on your use-case, QR-factorization could remove the need to invert a matrix. 

If you're trying to find `x` for `Ax = b`, if you have `A = QR`, where Q is  orthonormal and R is upper triangular, you can turn the problem into `Rx = transpose(Q)b`, which is easily solvable through back substitution. QR decomposition is being implemented in #265 and I think there'll be a pull request for `triangular_solve` coming soon.

The back substitution algorithm itself can be implemented as a placeholder while `triangular_solve` isn't available. If this can solve your problem, I might be able to help with it!

**Comment by ArthurHypersphere** [comment]
Nice draft implementation. I don't have a solid usecase yet for elixir, but I implemented a decent amount of trading related code in Elixir using my own stuff and after Nx was released I took it for a quick spin and noticed that I couldn't invert a square tensor, so wanted some additional color on this. I appreciate you all answering my inquiry. Don't want you guys to keep this open since you adequately answered the question. Is there a Discord or something that is active that I could join so that in the future questions like these won't clog up your Github issues?

Thanks, appreciate your help once again! 👍
